modules/users/server/controllers/users/users.authentication.server.controller.js

```diff
@@ -79,10 +79,32 @@ exports.signout = function(req, res) {
 };
 
 /**
+ * OAuth provider call
+ */
+exports.oauthCall = function(strategy, scope) {
+	return function(req, res, next) {
+		// Set redirection path on session.
+		// Do not redirect to a signin or signup page
+		if (!_.contains([
+			'/authentication/signin',
+			'/authentication/signup'
+		], req.query.redirect_to)) {
+			req.session.redirect_to = req.query.redirect_to;
+		}
+		// Authenticate
+		passport.authenticate(strategy, scope)(req, res, next);
+	};
+};
+
+/**
  * OAuth callback
  */
 exports.oauthCallback = function(strategy) {
 	return function(req, res, next) {
+		// Pop redirect URL from session
+		var sessionRedirectURL = req.session.redirect_to;
+		delete req.session.redirect_to;
```

Comment:
@ilanbiala , the callback is called by the provider. As I understand, it is not possible to pass the previous URL to the callback. This is why a `oauthCall` was introduced, so we have a way of storing the previous state. If we use `successRedirect`, then we would again need to pass the value for `successRedirect` somehow, perhaps through a property on session.

Here is where the idea came from [http://stackoverflow.com/a/13336055/2859410](http://stackoverflow.com/a/13336055/2859410)